samcli/commands/init/init_templates.py

```diff
@@ -62,55 +74,66 @@ def prompt_for_location(self, runtime, dependency_manager):
         raise InvalidInitTemplateError("Invalid template. This should not be possible, please raise an issue.")
 
     def location_from_app_template(self, runtime, dependency_manager, app_template):
         options = self.init_options(runtime, dependency_manager)
         try:
             template = next(item for item in options if self._check_app_template(item, app_template))
-            if template.get("init_location") is not None:
-                return template["init_location"]
-            if template.get("directory") is not None:
-                return os.path.join(self.repo_path, template["directory"])
+            if template.init_location is not None:
+                return template.init_location
+            if template.directory is not None:
+                return os.path.join(self.repo_path, template.directory)
             raise InvalidInitTemplateError("Invalid template. This should not be possible, please raise an issue.")
         except StopIteration as ex:
             msg = "Can't find application template " + app_template + " - check valid values in interactive init."
             raise InvalidInitTemplateError(msg) from ex
 
-    def _check_app_template(self, entry, app_template):
-        return entry["appTemplate"] == app_template
+    def _check_app_template(self, entry: TemplateOption, app_template):
+        return entry.app_template == app_template
 
-    def init_options(self, runtime, dependency_manager):
+    def init_options(self, runtime, dependency_manager) -> List[TemplateOption]:
         if self.clone_attempted is False:  # pylint: disable=compare-to-zero
             self._clone_repo()
         if self.repo_path is None:
             return self._init_options_from_bundle(runtime, dependency_manager)
         return self._init_options_from_manifest(runtime, dependency_manager)
 
-    def _init_options_from_manifest(self, runtime, dependency_manager):
+    def _init_options_from_manifest(self, runtime, dependency_manager) -> List[TemplateOption]:
         manifest_path = os.path.join(self.repo_path, "manifest.json")
         with open(str(manifest_path)) as fp:
             body = fp.read()
             manifest_body = json.loads(body)
-            templates = manifest_body.get(runtime)
-            if templates is None:
+            raw_templates = manifest_body.get(runtime)
+            if raw_templates is None:
                 # Fallback to bundled templates
                 return self._init_options_from_bundle(runtime, dependency_manager)
+            templates = [
+                TemplateOption(
+                    obj.get("directory", None),
+                    obj.get("displayName", None),
+                    obj.get("dependencyManager", None),
+                    obj.get("appTemplate", None),
+                    obj.get("initLocation", None),
+                    obj.get("isDynamicTemplate", None),
+                )
+                for obj in raw_templates
+            ]
```

Comment:
Copying to a typed object feels weird and wasteful; can we use [`TypedDict`](https://www.python.org/dev/peps/pep-0589/)s instead?